### PR TITLE
[lock] Provides default implementation when store does not supports the behavior

### PR DESCRIPTION
**Pull request by jderusse**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | yes
| Tickets       | /
| License       | MIT
| Doc PR        | /

All stores does not provide the same behaviors. Some are blocking, some allows shared Locks, ...
Issue is: When people use `lock` to use a behavior that is not supported by the store, they get an `UnsuportedException`, but they don't have any way to know if the store supports or not the behavior they want to use.

ie: when using the FrameworkBundle
```yaml
framework:
    lock: '%env(LOCK_DSN)%'
```
User (or bundle) can't use safely `$lock->acquire(true)` or `$lock->acquireRead()`. 

Given it's very easy to provide an fallback implementation to those behavior, this PR
- use fallback implementation when store does not support the behavior
- deprecate the RetryTillSaveStore (not needed anymore)

### CODE REVIEW

File: src/Symfony/Component/Lock/SharedLockStoreInterface.php [hunk 1]
```diff
@@ -22,15 +21,7 @@ interface SharedLockStoreInterface extends PersistingStoreInterface
     /**
      * Stores the resource if it's not locked for reading by someone else.
      *
-     * @throws NotSupportedException
```
Comment (jderusse) [inline src/Symfony/Component/Lock/SharedLockStoreInterface.php:25]: Is it a BC-break?

Comment (fabpot) [reply src/Symfony/Component/Lock/SharedLockStoreInterface.php:25]: I don't think so

File: src/Symfony/Component/Lock/Exception/NotSupportedException.php [hunk 2]
```diff
@@ -11,10 +11,14 @@
 
 namespace Symfony\Component\Lock\Exception;
 
+trigger_deprecation('symfony/lock', '5.1', '%s is deprecated, You should stop using it, as it will be removed in 6.0.', NotSupportedException::class);
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Lock/Exception/NotSupportedException.php:14]: 5.2

**Comment by nicolas-grekas** [comment]
Don't miss updating changelog+upgrade files.

File: src/Symfony/Component/Lock/Exception/NotSupportedException.php (hunk 2 above)
Comment (jderusse) [reply src/Symfony/Component/Lock/Exception/NotSupportedException.php]: thanks

File: src/Symfony/Component/Lock/SharedLockStoreInterface.php (hunk 1 above)
Comment (jderusse) [reply src/Symfony/Component/Lock/SharedLockStoreInterface.php:25]: nvm, this interface has been introduced in 5.2

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @jderusse.

**Pull request merged**